### ISSUE TITLE
ErrorException in SSH2.php line 3552: Undefined offset: 256

### ISSUE THREAD

**Issue by f4cklangzi** [phpseclib/phpseclib#1114]
in SSH2.php line 3552
at HandleExceptions->handleError('8', 'Undefined offset: 256', '/var/www/html/datingcloud-svn/vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php', '3552', array('client_channel' => '256', 'data' => 'myapp-1491973742/test.ipa')) in SSH2.php line 3552
at SSH2->_send_channel_packet('256', 'myapp-1491973742/test.ipa') in SFTP.php line 2789
at SFTP->_send_sftp_packet('16', 'myapp-1491973742/test.ipa') in SFTP.php line 633
at SFTP->_realpath('myapp-1491973742/test.ipa') in SFTP.php line 2296
at SFTP->file_exists('myapp-1491973742/test.ipa') in MakeController.php line 218

**Comment by f4cklangzi** [comment]
everytime i new sftp or ssh2 , did i need close it?how?

**Comment by f4cklangzi** [comment]
There is another problem，when i use sftp->get() a file,and every time just get a part of the file...

**Comment by terrafrost** [comment]
Here's where the channel is set to 256:

https://github.com/phpseclib/phpseclib/blob/2.0.4/phpseclib/Net/SFTP.php#L58

It's set when you login:

https://github.com/phpseclib/phpseclib/blob/2.0.4/phpseclib/Net/SFTP.php#L410

Nothing unsets the channel once it's been set. In particular, the following string does not occur in SSH2.php or SFTP.php:

```php
unset($this->channel_status
```

SFTP::file_exists calls stat:

https://github.com/phpseclib/phpseclib/blob/2.0.4/phpseclib/Net/SFTP.php#L2293

SFTP::stat does this:

https://github.com/phpseclib/phpseclib/blob/2.0.4/phpseclib/Net/SFTP.php#L1154

```php
        if (!($this->bitmap & SSH2::MASK_LOGIN)) {
            return false;
        }
``

If `$this->bitmap & SSH2::MASK_LOGIN` could be set whilst `_login` returned false that could explain how you're getting that error BUT...  I looked through SSH2.php...  `$this->bitmap |= self::MASK_LOGIN;` is only ever set immediately before `return true;`.

So, long story short, nothing sticks out to me in looking at the code. In lieu of that I'm going to need log files. You can get them by doing `define('NET_SSH2_LOGGING', 2);` at the top of your code and then do `$ssh->getLog();` after you encounter the error. And maybe share your code, as well.

**Comment by terrafrost** [comment]
> There is another problem，when i use sftp->get() a file,and every time just get a part of the file...

Logs. I need logs.

**Comment by f4cklangzi** [comment]
When i get a 10M file,My PHP code has been no response, so I can not get LOG

**Comment by terrafrost** [comment]
Well that's a new bit of information. Your last post said "*There is another problem，when i use sftp->get() a file,and every time just get a part of the file...*". Getting a part of a file doesn't necessarily mean that the PHP script is timing out.

In any event, you can still do logging but you'll have to set it up differently. Do `define('NET_SSH2_LOGGING', 4);` and `define('NET_SSH2_LOG_REALTIME_FILENAME', 'phpseclib.log');` at the top of the file. That'll enable real time logging to a file. By default that log file will include the most recent 1MB that was sent / received.

**Comment by f4cklangzi** [comment]
http://dating.qihetimes.com/phpseclib.log it's log,and the php is not response,Only get part of the file

**Comment by terrafrost** [comment]
I'm not seeing the Undefined offset error. And it looks like it's just chugging along, altho you are getting wildly varying speeds for the various incoming 32kb packets.

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0515, network: 0.0206s)
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.1127, network: 0.0925s)
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 2.7, network: 2.6625s)
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 29.1233, network: 29.0704s)
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 30.4887, network: 30.4708s)
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0268, network: 0s)
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.1349, network: 0.1172s)
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.1639, network: 0.1489s)
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0253, network: 0.0001s)
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.1385, network: 0.1063s)
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.1638, network: 0.1369s)
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.026, network: 0.0001s)
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0292, network: 0s)

So two of the packets took 30s to transfer 32kb. Two other packets took ~15s each. So that's 30s for 64kb or 1.5m for 128kb. But then there four 32kb packets with a transfer time of 0.1127s or less. That's 0.2s vs 1.5m. Why the difference? idk.

I mean, I guess I could see there being a delay like that if you were getting data from `/dev/random` but idk.

**Comment by salahy2** [comment]
hello, I am getting the same error, how did you solve it ??

**Comment by terrafrost** [comment]
@salahy2 - the fact that you responded to a closed ticket instead of opening a new one may mean subsequent replies may get lost. It might be beneficial for you to open up a new ticket.

Regardless...  can you post your SSH logs? You can get them by doing `define('NET_SSH2_LOGGING', 2);` at the top and `echo $ssh->getLog();` after the failed login attempt.

Also, post the exact message. You say you're getting the same error but you also said that in #1014. Maybe you really are getting the same error in both instances or maybe you're getting an error that looks very similar but has different line numbers. As is it's unclear.
